jgrapht-core/src/main/java/org/jgrapht/alg/clustering/ModularityMeasurer.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * (C) Copyright 2020-2020, by Dimitrios Michail and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * See the CONTRIBUTORS.md file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the
+ * GNU Lesser General Public License v2.1 or later
+ * which is available at
+ * http://www.gnu.org/licenses/old-licenses/lgpl-2.1-standalone.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR LGPL-2.1-or-later
+ */
+package org.jgrapht.alg.clustering;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.jgrapht.Graph;
+import org.jgrapht.GraphType;
+
+/**
+ * A <a href="https://en.wikipedia.org/wiki/Modularity_(networks)">modularity</a> measurer.
+ * 
+ * <p>
+ * This is a utility class which computes the modularity function. It takes as input a list of
+ * vertex classes $C$ and a graph $G$ and calculates: $Q = \frac{1}{2m} \sum_{ij} \left( A_{ij} -
+ * \frac{k_i k_j}{2m} \right) \delta(C_i, C_j)$. Here $m$ is the total number of edges and $k_i$ is
```

Comment:
So, as I understood the code, this formula depends on whether the graph is (a) weighted/unweighted (b) directed/undirected. 

a.1 If the graph is weighted, the `A_{i, j}` is the sum of edge weights between nodes `i` and `j`. `k_i` is the weight sum of the edges incident to the node `i`.
a.2 If the graph is unweighted, the `A_{i, j}` is the number of edges between nodes `i` and `j`. `k_i` is the number of edges incident to the node `i`.

b.1 If the graph is directed, the normalization factor in both cases is `1/m`. In this case, we distinguish between incoming edge weight sum and outcoming edge weight sum `k_i` and `k_j` (`k_i` means incoming sum, `k_j` means outgoing sum). The resulting formula is 
![image](https://user-images.githubusercontent.com/29632641/87225391-3f6ef080-c395-11ea-8c88-6f77b37b1f74.png)


b.2 If the graph is undirected, the normalization factor in both cases is `1/2m`. The resulting formula is 
![image](https://user-images.githubusercontent.com/29632641/87225345-e010e080-c394-11ea-921a-13875c796d86.png)

Comments:
* I would mention these cases in the class description as it's not clear that this class can be used for weighted/unweighted and directed/undirected graphs and that it returns different results depending in the graph type. 
* Why do we need an extra `2` [here](https://github.com/jgrapht/jgrapht/pull/967/files#diff-c942870efed40c5bd07eee6fe147295fR106)? This corresponds to this term in the formula above: 
![image](https://user-images.githubusercontent.com/29632641/87225463-dc318e00-c395-11ea-82a8-f5cf9d564583.png)